### PR TITLE
decision_proceduret API with assumptions

### PR DESCRIPTION
**Pull request by kroening**
- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by martin-cs — APPROVED** [review]
Seems like the sensible thing to do.

File: src/solvers/decision_procedure.cpp [hunk 1]
```diff
@@ -11,13 +11,21 @@ Author: Daniel Kroening, kroening@kroening.com
...
+
 decision_proceduret::~decision_proceduret()
 {
 }
 
 decision_proceduret::resultt decision_proceduret::operator()()
 {
-  return dec_solve();
+  return dec_solve(nil_exprt());
+}
```
Comment (tautschnig) [inline src/solvers/decision_procedure.cpp:23]: My preference would be to use std::optional at the interface rather than nil.

**Review by tautschnig — APPROVED** [review]

**Review by peterschrammel — APPROVED** [review]

File: src/solvers/decision_procedure.cpp (hunk 1 above)
Comment (kroening) [reply src/solvers/decision_procedure.cpp:23]: I was reluctant to do so since that can't avoid copying the expression.

Comment (tautschnig) [reply src/solvers/decision_procedure.cpp:23]: Wouldn't `const std::optional<exprt> &` do the trick?

Comment (kroening) [reply src/solvers/decision_procedure.cpp:23]: No, you'd still need to copy the expression into the optional when creating the optional.

Comment (kroening) [reply src/solvers/decision_procedure.cpp:23]: `std::optional< std::reference_wrapper<exprt> >` would work, but then it's getting more verbose.

**Pull request merged**